packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -749,6 +754,28 @@ export function resetComponentRoot(vm: VM) {
     vm.velements = EmptyArray;
 }
 
+// Helper function to traverse a tree of VFragment nodes and remove all root children.
+// This is moved into a separate function to minimize the perf/mem impact of the stack traversal
+// on non-vfragment use cases
+function removeFragmentChildren(vnode: VFragment, vm: VM) {
```

Comment:
Because VFragments can be nested inside other VFragments, we can't just remove the immediate children of the first fragment we find. We need to traverse the entire tree and grab all children that are immediate children of the root node.
```
<template>
    <div>Immediate child of root</div>
    VFragment(
        <div>Nested 1: also immediate child of root</div>
        VFragment(
            <div>Nested 2: also immediate child of root</div>
        )
    )
</template>
```
```
<x-custom-component>
    <div>Immediate child of root</div>
    <div>Nested 1: also immediate child of root</div>
    <div>Nested 2: also immediate child of root</div>
</x-custom-component>
```